packages/cardano-services/src/WsServer/server.ts

```diff
@@ -0,0 +1,278 @@
+// cSpell:ignore njson
+
+import { Cardano, WSMessage } from '@cardano-sdk/core';
+import { HandleRequestOptions, handleRequest } from './requests';
+import { Logger } from 'ts-log';
+import { NJSON } from 'next-json';
+import { Notification, Pool } from 'pg';
+import { Server, createServer } from 'http';
+import { WebSocket, WebSocketServer } from 'ws';
+import { initDB } from './db';
+import { v4 } from 'uuid';
+
+export interface WsServerConfiguration {
+  /** The heartbeat timeout in seconds. */
+  heartbeatTimeout?: number;
+
+  /** The port to listen. */
+  port: number;
+}
+
+export interface WsServerDependencies extends HandleRequestOptions {
+  /** The logger. */
+  logger: Logger;
+}
+
+// eslint-disable-next-line @typescript-eslint/no-empty-function
+const noop = () => {};
+
+export class CardanoWsServer {
+  private closeNotify: () => void;
+  private closing = false;
+  private db: Pool;
+  private logger: Logger;
+  private notifyConnected = false;
+  private server: Server;
+  private tip: Cardano.Tip;
+  private wss: WebSocketServer;
+
+  constructor(dependencies: WsServerDependencies, cfg: WsServerConfiguration) {
+    this.db = dependencies.db;
+    this.logger = dependencies.logger;
+
+    // Create the HTTP and the WebSocket servers
+    this.server = this.createHttpServer();
+    this.wss = new WebSocketServer({ server: this.server });
+
+    // Attach the handlers to the servers events
+
+    this.wss.on('connection', this.createOnConnection(dependencies, (cfg.heartbeatTimeout || 60) * 1000));
+
+    this.server.on('error', (error) => {
+      this.logger.error(error, 'Async error from HTTP server');
+      this.close();
+    });
+
+    this.wss.on('error', (error) => {
+      this.logger.error(error, 'Async error from WebSocket server');
+      this.close();
+    });
+
+    // Init the server
+    this.init(cfg.port).catch((error) => {
+      this.logger.error(error, 'Error in init sequence');
```

Comment:
It is not essential for a PoC, but we can fail all the services for a regular implementation.

Reply:
Sorry, could you please elaborate a bit? Thank you,

Reply:
If closing the WS causes service to shutdown, It covers what I was asking